threading: support for GGML_SCHED_PRIO_LOW, update thread info on Windows to avoid throttling

max-krasnyansky opened the pull request:
We talked about adding LOW priority for GGML threads in the original threadpool PR. It might be useful for some cases to avoid contention.

Latest Windows ARM64 releases started parking (offlining) the CPU cores more aggresively which results in suboptimal performance with n_threads > 4. To deal with that we now disable Power Throttling for our threads for the NORMAL and higher priorities.

*Make sure to read the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

slaren commented on ggml/src/ggml-cpu/ggml-cpu.c:
```diff
@@ -2417,12 +2417,30 @@ static bool ggml_thread_apply_priority(int32_t prio) {
+        p.ControlMask = THREAD_POWER_THROTTLING_EXECUTION_SPEED;
+        p.StateMask   = 0;
+
+        if (!SetThreadInformation(GetCurrentThread(), ThreadPowerThrottling, &p, sizeof(p))) {
```
Since this function is only available on Windows 8+, it would be nice to add a check here for the Windows version. `#if _WIN32_WINNT >= 0x0602` should do it.

max-krasnyansky committed:
threading: support for GGML_SCHED_PRIO_LOW, update thread info on Windows to avoid throttling

We talked about adding LOW priority for GGML threads in the original threadpool PR.
It might be useful for some cases to avoid contention.

Latest Windows ARM64 releases started parking (offlining) the CPU cores
more aggresively which results in suboptimal performance with n_threads > 4.
To deal with that we now disable Power Throttling for our threads for the NORMAL
and higher priorities.

Co-authored-by: Diego Devesa <slarengh@gmail.com>

max-krasnyansky committed:
threading: disable SetThreadInfo() calls for older Windows versions

slaren commented:
The CI failure here is a bit weird.. it is actually using mingw rather than MSVC, and the problem seems to be that this version of mingw does not support this API at all. We should just remove the msvc-arm64 build from the CI, along with the `arm64-windows-msvc.cmake` toolchain file, since it doesn't work, and this configuration is no longer supported.

slaren commented on tools/llama-bench/llama-bench.cpp:
```diff
@@ -315,7 +315,7 @@ static void print_usage(int /* argc */, char ** argv) {
     printf("  -r, --repetitions <n>                     number of times to repeat each test (default: %d)\n",
            cmd_params_defaults.reps);
-    printf("  --prio <0|1|2|3>                          process/thread priority (default: %d)\n",
+    printf("  --prio <-1,0|1|2|3>                          process/thread priority (default: %d)\n",
```
```suggestion
    printf("  --prio <-1|0|1|2|3>                          process/thread priority (default: %d)\n",
```

max-krasnyansky committed:
Update tools/llama-bench/llama-bench.cpp

Co-authored-by: Diego Devesa <slarengh@gmail.com>

MaggotHATE commented:
I'm getting compiling error after this PR. Windows 10, i7 8700, MinGW (w64devkit):

`ggml-cpu.c:2434:9: error: unknown type name 'THREAD_POWER_THROTTLING_STATE'; did you mean 'PPROCESS_POWER_THROTTLING_STATE'?`

Is there a check missing?

LostRuins commented:
`#if _WIN32_WINNT >= 0x0602`

If i am not wrong, this is set at compile time, which means that this will break windows 7 compatibility. Could it be changed to a runtime check instead?
